// earljs/dist/matchers/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.StringMatchingMatcher = exports.NumberCloseToMatcher = exports.ErrorMatcher = exports.ContainerWithMatcher = exports.Matcher = exports.ArrayWithMatcher = exports.AnythingMatcher = exports.AMatcher = void 0;
var A_1 = require("./A");
Object.defineProperty(exports, "AMatcher", { enumerable: true, get: function () { return A_1.AMatcher; } });
var Anything_1 = require("./Anything");
Object.defineProperty(exports, "AnythingMatcher", { enumerable: true, get: function () { return Anything_1.AnythingMatcher; } });
var ArrayWith_1 = require("./ArrayWith");
Object.defineProperty(exports, "ArrayWithMatcher", { enumerable: true, get: function () { return ArrayWith_1.ArrayWithMatcher; } });
var Base_1 = require("./Base");
Object.defineProperty(exports, "Matcher", { enumerable: true, get: function () { return Base_1.Matcher; } });
__exportStar(require("./booleans"), exports);
var ContainerWith_1 = require("./ContainerWith");
Object.defineProperty(exports, "ContainerWithMatcher", { enumerable: true, get: function () { return ContainerWith_1.ContainerWithMatcher; } });
var Error_1 = require("./Error");
Object.defineProperty(exports, "ErrorMatcher", { enumerable: true, get: function () { return Error_1.ErrorMatcher; } });
var NumberCloseTo_1 = require("./NumberCloseTo");
Object.defineProperty(exports, "NumberCloseToMatcher", { enumerable: true, get: function () { return NumberCloseTo_1.NumberCloseToMatcher; } });
__exportStar(require("./numbers"), exports);
__exportStar(require("./optionals"), exports);
var StringMatching_1 = require("./StringMatching");
Object.defineProperty(exports, "StringMatchingMatcher", { enumerable: true, get: function () { return StringMatching_1.StringMatchingMatcher; } });
